cannon/mipsevm/mips_syscalls.go

```diff
@@ -0,0 +1,195 @@
+)
+
+const (
+	sysMmap      = 4090
+	sysBrk       = 4045
+	sysClone     = 4120
+	sysExitGroup = 4246
+	sysRead      = 4003
+	sysWrite     = 4004
+	sysFcntl     = 4055
+)
+
+const (
+	fdStdin         = 0
+	fdStdout        = 1
+	fdStderr        = 2
+	fdHintRead      = 3
+	fdHintWrite     = 4
+	fdPreimageRead  = 5
+	fdPreimageWrite = 6
+)
+
+const (
+	MipsEBADF  = 0x9
+	MipsEINVAL = 0x16
+)
+
+type PreimageReader func(key [32]byte, offset uint32) (dat [32]byte, datLen uint32)
+type MemTracker func(addr uint32)
+
+func getSyscallArgs(registers *[32]uint32) (syscallNum, a0, a1, a2 uint32) {
+	syscallNum = registers[2] // v0
+
+	a0 = registers[4]
+	a1 = registers[5]
+	a2 = registers[6]
+
+	return syscallNum, a0, a1, a2
+}
+
+func handleSysMmap(a0, a1, heap uint32) (v0, v1, newHeap uint32) {
+	v1 = uint32(0)
+	newHeap = heap
+
+	sz := a1
+	if sz&PageAddrMask != 0 { // adjust size to align with page size
+		sz += PageSize - (sz & PageAddrMask)
+	}
+	if a0 == 0 {
+		v0 = heap
+		//fmt.Printf("mmap heap 0x%x size 0x%x\n", v0, sz)
+		newHeap += sz
+	} else {
+		v0 = a0
+		//fmt.Printf("mmap hint 0x%x size 0x%x\n", v0, sz)
+	}
+
+	return v0, v1, newHeap
+}
+
+func handleSysRead(a0, a1, a2 uint32, preimageKey [32]byte, preimageOffset uint32, preimageReader PreimageReader, memory *Memory, memTracker MemTracker) (v0, v1, newPreimageOffset uint32) {
+	// args: a0 = fd, a1 = addr, a2 = count
+	// returns: v0 = read, v1 = err code
+	v0 = uint32(0)
+	v1 = uint32(0)
+	newPreimageOffset = preimageOffset
+
+	switch a0 {
+	case fdStdin:
+		// leave v0 and v1 zero: read nothing, no error
+	case fdPreimageRead: // pre-image oracle
+		effAddr := a1 & 0xFFffFFfc
+		memTracker(effAddr)
+		mem := memory.GetMemory(effAddr)
+		dat, datLen := preimageReader(preimageKey, preimageOffset)
+		//fmt.Printf("reading pre-image data: addr: %08x, offset: %d, datLen: %d, data: %x, key: %s  count: %d\n", a1, m.state.PreimageOffset, datLen, dat[:datLen], m.state.PreimageKey, a2)
+		alignment := a1 & 3
+		space := 4 - alignment
+		if space < datLen {
+			datLen = space
+		}
+		if a2 < datLen {
+			datLen = a2
+		}
+		var outMem [4]byte
+		binary.BigEndian.PutUint32(outMem[:], mem)
+		copy(outMem[alignment:], dat[:datLen])
+		memory.SetMemory(effAddr, binary.BigEndian.Uint32(outMem[:]))
+		newPreimageOffset += datLen
+		v0 = datLen
+		//fmt.Printf("read %d pre-image bytes, new offset: %d, eff addr: %08x mem: %08x\n", datLen, m.state.PreimageOffset, effAddr, outMem)
+	case fdHintRead: // hint response
+		// don't actually read into memory, just say we read it all, we ignore the result anyway
+		v0 = a2
+	default:
+		v0 = 0xFFffFFff
+		v1 = MipsEBADF
+	}
+
+	return v0, v1, newPreimageOffset
+}
+
+func handleSysWrite(a0, a1, a2 uint32, lastHint hexutil.Bytes, preimageKey [32]byte, preimageOffset uint32, oracle PreimageOracle, memory *Memory, memTracker MemTracker, stdOut, stdErr io.Writer) (v0, v1 uint32, newLastHint hexutil.Bytes, newPreimageKey common.Hash, newPreimageOffset uint32) {
```

Comment:
Some of these function signatures are kind of crazy, but not sure how to avoid it... 🤷‍♀️

Reply:
Yeah. I think it's fine though because there is only one callsite.